evap/rewards/tools.py

```diff
@@ -59,30 +59,45 @@ def reward_points_of_user(user):
 def is_semester_activated(semester):
     return SemesterActivation.objects.filter(semester=semester, is_active=True).exists()
 
-
-# Signal handlers
-
-@receiver(Course.course_evaluated)
-def grant_reward_points(sender, **kwargs):
+def grant_reward_points(user, semester):
     # grant reward points if all conditions are fulfilled
 
-    request = kwargs['request']
-    semester = kwargs['semester']
-    if not can_user_use_reward_points(request.user):
-        return
+    if not can_user_use_reward_points(user):
+        return False
     # has the semester been activated for reward points?
     if not is_semester_activated(semester):
-        return
+        return False
     # does the user have at least one required course in this semester?
-    required_courses = Course.objects.filter(participants=request.user, semester=semester, is_required_for_reward=True)
+    required_courses = Course.objects.filter(participants=user, semester=semester, is_required_for_reward=True)
     if not required_courses.exists():
-        return
+        return False
     # does the user not participate in any more required courses in this semester?
-    if required_courses.exclude(voters=request.user).exists():
-        return
+    if required_courses.exclude(voters=user).exists():
+        return False
     # did the user not already get reward points for this semester?
-    if RewardPointGranting.objects.filter(user_profile=request.user, semester=semester).exists():
-        return
+    if RewardPointGranting.objects.filter(user_profile=user, semester=semester).exists():
+        return False
     # grant reward points
-    RewardPointGranting.objects.create(user_profile=request.user, semester=semester, value=settings.REWARD_POINTS_PER_SEMESTER)
-    messages.success(request, _("You just have earned reward points for this semester because you evaluated all your courses. Thank you very much!"))
+    RewardPointGranting.objects.create(user_profile=user, semester=semester, value=settings.REWARD_POINTS_PER_SEMESTER)
+    return True
+
+# Signal handlers
+
+@receiver(Course.course_evaluated)
+def grant_reward_points_after_evaluate(sender, **kwargs):
+    request = kwargs['request']
+    semester = kwargs['semester']
+
+    if grant_reward_points(request.user, semester):
+        messages.success(request, _("You just have earned reward points for this semester because you evaluated all your courses. Thank you very much!"))
+
+@receiver(models.signals.m2m_changed, sender=Course.participants.through)
+def grant_reward_points_after_delete(instance, action, reverse, pk_set, **kwargs):
+    # if users do not need to evaluate a course anymore, they may have earned reward points
+    if action == "post_remove":
```

Comment:
we have a loose rule which we didn't tell you yet that in python code user-facing strings use double quotes, and other strings single quotes. this would be one of those other strings :)